### PR TITLE
sql: add an internal table showing lost table descriptors

### LINKED ISSUE #62800: sql: add an internal table showing lost table descriptors
As a part of the follow on work for https://github.com/cockroachdb/cockroach/issues/62532 we should have an internal table to capture table descriptor ID's with leftover data

### PR DESCRIPTION
**Pull request by fqazi**
Fixes: #62800
Previously, there was no way of knowing if any data was left
over by table descriptors that were force delete via internal
functions. This was problematic, since these tables could
still have data left over in storage, even though the descriptors
were cleaned up. We added a builtin to clear this data, but we
have no way of knowing which descriptors have leaked data. To
address this a new internal table
crdb_internal.lost_descriptors_with_data.

Release note (sql change): Added crdb_internal.lost_descriptors_with_data
to show descriptors that have no entries but data left behind.

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/schema_repair [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+					AS descriptor
+			FROM
+				system.descriptor
+		)
+SELECT
+	descriptor
+FROM
+	descs
+WHERE
+	id = $t_id;
```
Comment (postamar) [inline pkg/sql/logictest/testdata/logic_test/schema_repair:41]: Can you reformat this through https://sqlfum.pt/ please?

**Review by postamar — COMMENTED** [review]
That's it for me for now.

File: pkg/sql/logictest/testdata/logic_test/schema_repair (hunk 1 above)
Comment (fqazi) [reply pkg/sql/logictest/testdata/logic_test/schema_repair:41]: That actually went through sqlfum.pt already. Passed it again and see no difference.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62801)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/cli/zip_test.go, line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MX3dIm7i5A-axkZnfZ6-r1-76:-MXbcDpS4tuMPhXYzAV1:b-4sb5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc4544592892ae0ac8b2f76e519ba9a6173940e9/pkg/cli/zip_test.go#L76)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: something about this name bugs me, but I don't know what or why, and I'm not able to come up with a more suitable alternative. Sorry this isn't a super constructive comment! 
</blockquote></details>

Yes, I'm not a fan of the name either. Maybe @ajwerner has a different idea?
___
*[pkg/sql/crdb_internal.go, line 4402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MX2ncTY3utUkK3HWUUa-r1-4402:-MXbcL0H1cVXazWtuL7O:besit05) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc4544592892ae0ac8b2f76e519ba9a6173940e9/pkg/sql/crdb_internal.go#L4402)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is there a specific reason you need to pass `AvoidCached: true`? 

In any case I suggest you get all descriptors with `p.Descriptors().GetAllDescriptorsUnvalidated(...)`, because as things are if a descriptor is invalid, an error is returned causing to exit the loop. I don't think that that's the desired behaviour. Perhaps you could add a test that covers this case? Have a corrupted descriptor at a certain id and data-with-missing-descriptor at an other, greater id, and check that the latter id shows up in the  `lost_descriptors_with_data` resultset.

</blockquote></details>

Done. Moved over to GetAllDescriptorsUnvalidated API
___
*[pkg/sql/crdb_internal.go, line 4403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MX2ncTY3utUkK3HWUUa-r1-4403:-MXbcR5V1ktumRssvIa5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc4544592892ae0ac8b2f76e519ba9a6173940e9/pkg/sql/crdb_internal.go#L4403)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Conversely, `Required: false` should imply that `err` won't ever be `ErrDescriptorNotFound`, I believe.
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 4417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MX2ncTY3utUkK3HWUUa-r1-4417:-MXbcRwE8Gjqic0fs9Wt:bovdw9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc4544592892ae0ac8b2f76e519ba9a6173940e9/pkg/sql/crdb_internal.go#L4417)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I know very little about the kv api so I'll defer to others as to whether this request is as it should be. Personally I'm curious as to whether it's preferable to set `MaxSpanRequestKeys` or `TargetBytes` in this use case. Also, out of curiosity, what's the rationale behind `ScanFormat = BATCH_RESPONSE` if you don't mind sharing?
</blockquote></details>

The ScanFormat was based on other similar code, we aren't going over the results, so it won't make a difference (maybe the normal format will be cheaper?). But, as far as I can tell the layout will be interleaved batches of keys followed by values, instead of just a slice of key values. 

I'll let @ajwerner confirm which is better, I went with MaxSpanRequests just based on your suggestion from earlier.

File: pkg/sql/logictest/testdata/logic_test/schema_repair (hunk 1 above)
Comment (postamar) [reply pkg/sql/logictest/testdata/logic_test/schema_repair:41]: You're right, sorry about that. I think I messed with the defaults in sqlfumpt. Whoops.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62801)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/logictest/testdata/logic_test/schema_repair, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MXD8oymdX5HCG2wU2ZT-r1-41:-MXcFSSa7AkLNNI7bbM-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc4544592892ae0ac8b2f76e519ba9a6173940e9/pkg/sql/logictest/testdata/logic_test/schema_repair#L41)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You're right, sorry about that. I think I messed with the defaults in sqlfumpt. Whoops.
</blockquote></details>

Done.

File: pkg/sql/crdb_internal.go [hunk 2]
```diff
@@ -4379,3 +4380,61 @@ CREATE TABLE crdb_internal.cross_db_references (
...
+		if err != nil {
+			return err
+		}
+		// Get all descriptors which will be used to determine
+		// which ones are missing
+		descriptors, err := catalogkv.GetAllDescriptorsUnvalidated(ctx, p.txn, p.extendedEvalCtx.Codec)
+		if err != nil {
+			return err
+		}
+		descriptorMap := make(map[descpb.ID]*catalog.Descriptor)
```
Comment (postamar) [inline pkg/sql/crdb_internal.go:4406]: `catalog.Descriptor` is an interface type and so is already effectively a pointer.

File: pkg/sql/crdb_internal.go [hunk 3]
```diff
@@ -4379,3 +4380,61 @@ CREATE TABLE crdb_internal.cross_db_references (
...
+		// Loop over every possible descriptor ID
+		for id := int64(1); id < maxDescID; id++ {
+			// Skip over descriptors that are known
+			if _, ok := descriptorMap[descpb.ID(id)]; ok {
+				continue
+			}
+			// Generate a scan over the key space of the table to see
+			// if any data exists.
+			b := kv.Batch{}
+			prefix := p.extendedEvalCtx.Codec.TablePrefix(uint32(id))
```
Comment (postamar) [inline pkg/sql/crdb_internal.go:4419]: This didn't occur to me earlier but why not make `id` of type `uint32`? We don't really support IDs beyond 2^32 anyway.

**Review by postamar — COMMENTED** [review]
Just a few little nits on my end, no further complaints, I'm otherwise happy with this. As discussed earlier the kv api call will need to be reviewed by someone else.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62801)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/crdb_internal.go, line 4406 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MX2ncTY3utUkK3HWUUa-r2-4406:-MXczzRK6Fs5PQNgU9PN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d23b64a291739ddf268e1c5e55ab94b001931f/pkg/sql/crdb_internal.go#L4406)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`catalog.Descriptor` is an interface type and so is already effectively a pointer.
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 4419 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MX2ncTY3utUkK3HWUUa-r2-4419:-MXczzwr0cGruKEDYg1W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8d23b64a291739ddf268e1c5e55ab94b001931f/pkg/sql/crdb_internal.go#L4419)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This didn't occur to me earlier but why not make `id` of type `uint32`? We don't really support IDs beyond 2^32 anyway.
</blockquote></details>

Done.

**Review by ajwerner — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62801)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/crdb_internal.go, line 4392 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MXec__s8dKVDYVh9UCR:-MXec__s8dKVDYVh9UCS:btptdwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/29e21ba74ad82d883437caffa28137b020c77d7a/pkg/sql/crdb_internal.go#L4392)):*
> ```Go
> );`,
> 	populate: func(ctx context.Context, p *planner, dbContext *dbdesc.Immutable, addRow func(...tree.Datum) error) error {
> 		maxDescIDKeyVal, err := p.extendedEvalCtx.DB.Get(context.Background(), p.extendedEvalCtx.Codec.DescIDSequenceKey())
> ```

This implementation feels unsatisfyingly inefficient. We have two sorted sets here that we're dealing with.  The machinery you are looking for on the kv side is https://github.com/cockroachdb/cockroach/blob/ac572af5be4e35dbdd58a95e32e5d1f6c24bc24c/pkg/kv/kvclient/kvcoord/range_iter.go#L39.

In principle you should be able to just alternate stepping through/seeking the two iterators and accumulating some scan requests. My guess is that this whole thing can, in many or most cases be turned into a much much smaller number of scans. For the scans, I'd prefer you build a batch with a request corresponding to each gap you expect to be empty. At the end, you can send the batch and for each span that comes back non-empty, you can go iterate through the possibilities. This approach should make this call roughly O(1) round-trips in the case where the range cache is warm. 

For the descriptors, I think it's ~fine to do what you're doing but I don't really see any benefit over just using the collection to get all of the descriptors. Given we buffer them all in RAM all over the place, for now this is fine.

**Review by ajwerner — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62801)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/crdb_internal.go, line 4392 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MXec__s8dKVDYVh9UCR:-MXefqzu-JGkQnF_iaFg:b-3c4epu) ([raw file](https://github.com/cockroachdb/cockroach/blob/29e21ba74ad82d883437caffa28137b020c77d7a/pkg/sql/crdb_internal.go#L4392)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This implementation feels unsatisfyingly inefficient. We have two sorted sets here that we're dealing with.  The machinery you are looking for on the kv side is https://github.com/cockroachdb/cockroach/blob/ac572af5be4e35dbdd58a95e32e5d1f6c24bc24c/pkg/kv/kvclient/kvcoord/range_iter.go#L39.

In principle you should be able to just alternate stepping through/seeking the two iterators and accumulating some scan requests. My guess is that this whole thing can, in many or most cases be turned into a much much smaller number of scans. For the scans, I'd prefer you build a batch with a request corresponding to each gap you expect to be empty. At the end, you can send the batch and for each span that comes back non-empty, you can go iterate through the possibilities. This approach should make this call roughly O(1) round-trips in the case where the range cache is warm. 

For the descriptors, I think it's ~fine to do what you're doing but I don't really see any benefit over just using the collection to get all of the descriptors. Given we buffer them all in RAM all over the place, for now this is fine.


</blockquote></details>

Having brought up the RangeIterator, I'm now questioning myself. It's only really useful if you wanted to look at split points and make assumptions about how that all works. Let's not. I guess this comment just collapses to, please build batches of scans over spans you expect to be empty and then, only if they come back non-empty, should you go try to understand why. I anticipate that we will commonly see very large descriptor gaps.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62801)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/crdb_internal.go, line 4392 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MXec__s8dKVDYVh9UCR:-MXhi5u--apom_mmq3c0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/29e21ba74ad82d883437caffa28137b020c77d7a/pkg/sql/crdb_internal.go#L4392)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Having brought up the RangeIterator, I'm now questioning myself. It's only really useful if you wanted to look at split points and make assumptions about how that all works. Let's not. I guess this comment just collapses to, please build batches of scans over spans you expect to be empty and then, only if they come back non-empty, should you go try to understand why. I anticipate that we will commonly see very large descriptor gaps. 
</blockquote></details>

Done.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62801)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/crdb_internal.go, line 4401 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MYetkPQEhTJtoAP0c1B:-MYetkPR1QOnNdLqili2:b-lpx8dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/40cdaf73ede44b785875b4c371c13f71fcb0731d/pkg/sql/crdb_internal.go#L4401)):*
> ```Go
> 		}
> 		// Get all descriptors which will be used to determine
> 		// which ones are missing
> ```

nit: period. 
___
*[pkg/sql/crdb_internal.go, line 4423 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MYeuPYyAIzE-tXnDuF1:-MYeuPYyAIzE-tXnDuF2:b-74dxiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/40cdaf73ede44b785875b4c371c13f71fcb0731d/pkg/sql/crdb_internal.go#L4423)):*
> ```Go
> 			b.Header.MaxSpanRequestKeys = 1
> 			scanRequest := roachpb.NewScan(unusedDescSpan.Key, unusedDescSpan.EndKey, false).(*roachpb.ScanRequest)
> 			scanRequest.ScanFormat = roachpb.BATCH_RESPONSE
> ```

Please set a limit on the initial scan. 
___
*[pkg/sql/crdb_internal.go, line 4462 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MYeu6F51-wNWpj3sa-P:-MYeu6F51-wNWpj3sa-Q:b-opaquf) ([raw file](https://github.com/cockroachdb/cockroach/blob/40cdaf73ede44b785875b4c371c13f71fcb0731d/pkg/sql/crdb_internal.go#L4462)):*
> ```Go
> 		}
> 		// Loop over every possible descriptor ID
> 		for id := 1; id < int(maxDescID); id++ {
> ```

This still seems to be O(gaps). I'd think you can accumulate scans in a batch and then only run a batch once you've accumulated some large number of scan requests. Lots of things create gaps. Temp tables especially. 
___
*[pkg/sql/crdb_internal.go, line 4462 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MYeunGcBl-KKB6Xo5JV:-MYeunGcBl-KKB6Xo5JW:b8hmyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/40cdaf73ede44b785875b4c371c13f71fcb0731d/pkg/sql/crdb_internal.go#L4462)):*
> ```Go
> 		}
> 		// Loop over every possible descriptor ID
> 		for id := 1; id < int(maxDescID); id++ {
> ```

Start at `keys.MinUserDescID`

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62801)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

___
*[pkg/sql/crdb_internal.go, line 4423 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MYeuPYyAIzE-tXnDuF1:-MYfRSMJ4s792w5JN5j1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/40cdaf73ede44b785875b4c371c13f71fcb0731d/pkg/sql/crdb_internal.go#L4423)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Please set a limit on the initial scan. 
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 4462 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-MYeu6F51-wNWpj3sa-P:-MYfRLt7090qWPqN_D5h:b-g9244w) ([raw file](https://github.com/cockroachdb/cockroach/blob/40cdaf73ede44b785875b4c371c13f71fcb0731d/pkg/sql/crdb_internal.go#L4462)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This still seems to be O(gaps). I'd think you can accumulate scans in a batch and then only run a batch once you've accumulated some large number of scan requests. Lots of things create gaps. Temp tables especially. 
</blockquote></details>

Scans will be run serially as you pointed on Slack with the limit, so disregarding this since there will be no gains.

**Review by ajwerner — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-:-MYfUYQ63N9WCb9a2nC4:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)

**Comment by fqazi** [comment]
> :lgtm:
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62801#-:-MYfUYQ63N9WCb9a2nC4:bnfp4nl)_ status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)

Thanks for the review!

**Comment by fqazi** [comment]
bors r=ajwerner

**Comment by fqazi** [comment]
bors r=ajwerner

**Pull request merged**